File: packages/webapp/src/core/base64-mentions.ts
import { normalizeBase64 } from '@slicc/shared-ts';

export const MIN_PAYLOAD_CHARS = 128;

export interface Base64Candidate {
  raw: string;

  data: string;

  declaredMime?: string;

  start: number;

  end: number;
}

const DATA_URL_RE = /data:([\w.+-]+\/[\w.+-]+)(?:;[\w.+-]+=[^;,]*)*;base64,([A-Za-z0-9+/=]+)/g;

const RUN_OPENERS = '\\s"\'`(\\[{<,;:=';
const RUN_CLOSERS = '\\s"\'`)\\]}>,;:.!?';

const BARE_RUN_RE = new RegExp(
  `(?:^|[${RUN_OPENERS}])([A-Za-z0-9+/]{${MIN_PAYLOAD_CHARS},}={0,2})(?=$|[${RUN_CLOSERS}])`,
  'g'
);

const MIN_WRAP_COLUMNS = 16;

interface SourceLine {
  start: number;
  end: number;
  text: string;
}

function scanLines(text: string): SourceLine[] {
  const lines: SourceLine[] = [];
  let start = 0;
  for (;;) {
    const br = text.indexOf('\n', start);
    const end = br < 0 ? text.length : br;

    const trimmed = text[end - 1] === '\r' ? end - 1 : end;
    lines.push({ start, end: trimmed, text: text.slice(start, trimmed) });
    if (br < 0) return lines;
    start = br + 1;
  }
}

const PURE_ALPHABET_RE = /^[A-Za-z0-9+/]+$/;
const PADDED_TAIL_RE = /^[A-Za-z0-9+/]+={0,2}$/;

function findWrappedBlocks(text: string): Base64Candidate[] {
  const lines = scanLines(text);
  const blocks: Base64Candidate[] = [];

  for (let i = 0; i < lines.length; i += 1) {
    const width = wrapWidthAt(lines, i);
    if (width === null) continue;
    const tail = blockEnd(lines, i, width);
    if (tail === i) continue;
    const pieces = lines.slice(i, tail + 1);
    const lead = precedingFragment(lines, i, width);
    if (lead) pieces.unshift(lead);
    const block = claimBlock(text, pieces);
    if (block) blocks.push(block);
    i = tail;
  }

  return blocks;
}

function wrapWidthAt(lines: readonly SourceLine[], i: number): number | null {
  const first = lines[i];
  if (!first) return null;
  const width = first.text.length;
  if (width < MIN_WRAP_COLUMNS || width % 4 !== 0) return null;
  return PURE_ALPHABET_RE.test(first.text) ? width : null;
}

function blockEnd(lines: readonly SourceLine[], i: number, width: number): number {
  let last = i;
  for (let j = i + 1; j < lines.length; j += 1) {
    const line = lines[j];
    if (!line || line.text.length !== width || !PURE_ALPHABET_RE.test(line.text)) break;
    last = j;
  }
  const next = lines[last + 1];
  if (!next || next.text.length === 0 || next.text.length >= width) return last;
  if (!PADDED_TAIL_RE.test(next.text)) return last;
  const after = lines[last + 2];
  const blockContinues =
    after !== undefined && after.text.length === width && PURE_ALPHABET_RE.test(after.text);
  return blockContinues ? last : last + 1;
}

function precedingFragment(
  lines: readonly SourceLine[],
  i: number,
  width: number
): SourceLine | null {
  const prev = lines[i - 1];
  if (!prev || prev.text.length <= width) return null;
  const cut = prev.text.length - width;
  const suffix = prev.text.slice(cut);
  if (!PURE_ALPHABET_RE.test(suffix)) return null;
  if (PURE_ALPHABET_RE.test(prev.text[cut - 1] ?? ' ')) return null;
  return { start: prev.start + cut, end: prev.end, text: suffix };
}

function claimBlock(text: string, block: readonly SourceLine[]): Base64Candidate | null {
  const joined = block.map((line) => line.text).join('');
  if (joined.replace(/=+$/, '').length < MIN_PAYLOAD_CHARS) return null;
  if (joined.length % 4 !== 0) return null;
  const data = normalizeBase64(joined);
  if (!data) return null;
  const start = block[0]?.start ?? 0;
  const end = block[block.length - 1]?.end ?? start;
  return { raw: text.slice(start, end), data, start, end };
}

export function findBase64Mentions(text: string): Base64Candidate[] {
  const found: Base64Candidate[] = [];
  const claimed: Array<[number, number]> = [];

  const overlaps = (start: number, end: number): boolean =>
    claimed.some(([s, e]) => start < e && end > s);

  DATA_URL_RE.lastIndex = 0;
  for (const match of text.matchAll(DATA_URL_RE)) {
    const payload = match[2] ?? '';
    if (payload.length < MIN_PAYLOAD_CHARS) continue;
    const data = normalizeBase64(payload);
    if (!data) continue;
    const start = match.index ?? 0;
    const end = start + match[0].length;
    found.push({ raw: match[0], data, declaredMime: match[1] ?? '', start, end });
    claimed.push([start, end]);
  }

  for (const block of findWrappedBlocks(text)) {
    if (overlaps(block.start, block.end)) continue;
    found.push(block);
    claimed.push([block.start, block.end]);
  }

  BARE_RUN_RE.lastIndex = 0;
  for (const match of text.matchAll(BARE_RUN_RE)) {
    const raw = match[1] ?? '';

    const start = (match.index ?? 0) + (match[0].length - raw.length);
    const end = start + raw.length;
    if (overlaps(start, end)) continue;

    if (raw.length % 4 !== 0) continue;
    const data = normalizeBase64(raw);
    if (!data) continue;
    found.push({ raw, data, start, end });
    claimed.push([start, end]);
  }

  return found.sort((a, b) => a.start - b.start);
}

File: packages/ios-app/SliccFollower/Models/Base64Mentions.swift
import Foundation

enum Base64Mentions {

    struct Candidate: Equatable {

        let data: String

        let declaredMime: String?

        let range: Range<String.Index>
    }

    static let minimumPayloadCharacters = 128

    private static let minimumWrapColumns = 16

    private static let dataURLRegex: NSRegularExpression? = {
        try? NSRegularExpression(
            pattern: #"data:([\w.+-]+/[\w.+-]+)(?:;[\w.+-]+=[^;,]*)*;base64,([A-Za-z0-9+/=]+)"#)
    }()

    private static let bareRunRegex: NSRegularExpression? = {
        let openers = #"\s"'`(\[{<,;:="#
        let closers = #"\s"'`)\]}>,;:.!?"#
        return try? NSRegularExpression(
            pattern:
                "(?:^|[\(openers)])([A-Za-z0-9+/]{\(minimumPayloadCharacters),}={0,2})(?=$|[\(closers)])"
        )
    }()

    static func scan(_ text: String) -> [Candidate] {
        guard text.count >= minimumPayloadCharacters else { return [] }
        var found: [Candidate] = []
        var claimed: [Range<String.Index>] = []

        if let regex = dataURLRegex {
            for match in regex.matches(in: text, range: NSRange(text.startIndex..., in: text)) {
                guard let whole = Range(match.range, in: text),
                    let payloadRange = Range(match.range(at: 2), in: text),
                    let mimeRange = Range(match.range(at: 1), in: text)
                else { continue }
                let payload = String(text[payloadRange])
                guard payload.count >= minimumPayloadCharacters,
                    let data = normalized(payload)
                else { continue }
                found.append(
                    Candidate(data: data, declaredMime: String(text[mimeRange]), range: whole))
                claimed.append(whole)
            }
        }

        for block in wrappedBlocks(in: text) where !claimed.contains(where: { $0.overlaps(block.range) }) {
            found.append(block)
            claimed.append(block.range)
        }

        if let regex = bareRunRegex {
            for match in regex.matches(in: text, range: NSRange(text.startIndex..., in: text)) {
                guard let run = Range(match.range(at: 1), in: text) else { continue }
                if claimed.contains(where: { $0.overlaps(run) }) { continue }
                let raw = String(text[run])

                guard raw.count % 4 == 0, let data = normalized(raw) else { continue }
                found.append(Candidate(data: data, declaredMime: nil, range: run))
                claimed.append(run)
            }
        }

        found.sort { $0.range.lowerBound < $1.range.lowerBound }
        return found
    }

    private struct SourceLine {
        let range: Range<String.Index>
        let text: Substring
    }

    private static func wrappedBlocks(in text: String) -> [Candidate] {
        let lines = scanLines(text)
        var blocks: [Candidate] = []
        var index = 0
        while index < lines.count {
            guard let width = wrapWidth(lines, at: index) else {
                index += 1
                continue
            }
            let tail = blockEnd(lines, from: index, width: width)
            if tail == index {
                index += 1
                continue
            }
            var pieces = Array(lines[index...tail])
            if let lead = precedingFragment(lines, at: index, width: width) {
                pieces.insert(lead, at: 0)
            }
            if let block = claim(pieces) { blocks.append(block) }
            index = tail + 1
        }
        return blocks
    }

    private static func scanLines(_ text: String) -> [SourceLine] {
        var lines: [SourceLine] = []
        var start = text.startIndex
        while true {
            let brk = text[start...].firstIndex(of: "\n")
            let end = brk ?? text.endIndex

            var trimmed = end
            if trimmed > start, text[text.index(before: trimmed)] == "\r" {
                trimmed = text.index(before: trimmed)
            }
            lines.append(SourceLine(range: start..<trimmed, text: text[start..<trimmed]))
            guard let brk else { return lines }
            start = text.index(after: brk)
        }
    }

    private static func isPureAlphabet(_ s: Substring) -> Bool {
        !s.isEmpty && s.allSatisfy { $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "+" || $0 == "/") }
    }

    private static func isPaddedTail(_ s: Substring) -> Bool {
        let padding = s.drop(while: { $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "+" || $0 == "/") })
        return s.count > padding.count && padding.count <= 2 && padding.allSatisfy { $0 == "=" }
    }

    private static func wrapWidth(_ lines: [SourceLine], at index: Int) -> Int? {
        let first = lines[index]
        let width = first.text.count
        guard width >= minimumWrapColumns, width % 4 == 0, isPureAlphabet(first.text) else {
            return nil
        }
        return width
    }

    private static func blockEnd(_ lines: [SourceLine], from index: Int, width: Int) -> Int {
        var last = index
        var j = index + 1
        while j < lines.count, lines[j].text.count == width, isPureAlphabet(lines[j].text) {
            last = j
            j += 1
        }
        guard last + 1 < lines.count else { return last }
        let next = lines[last + 1].text
        guard !next.isEmpty, next.count < width, isPaddedTail(next) else { return last }
        if last + 2 < lines.count, lines[last + 2].text.count == width,
            isPureAlphabet(lines[last + 2].text)
        {
            return last
        }
        return last + 1
    }

    private static func precedingFragment(_ lines: [SourceLine], at index: Int, width: Int)
        -> SourceLine?
    {
        guard index > 0 else { return nil }
        let prev = lines[index - 1]
        guard prev.text.count > width else { return nil }
        let cut = prev.text.index(prev.text.endIndex, offsetBy: -width)
        let suffix = prev.text[cut...]
        guard isPureAlphabet(suffix) else { return nil }
        let before = prev.text[prev.text.index(before: cut)]
        guard !isPureAlphabet(Substring(String(before))) else { return nil }
        return SourceLine(range: cut..<prev.range.upperBound, text: suffix)
    }

    private static func claim(_ block: [SourceLine]) -> Candidate? {
        let joined = block.map { String($0.text) }.joined()
        let unpadded = joined.reversed().drop(while: { $0 == "=" }).count
        guard unpadded >= minimumPayloadCharacters, joined.count % 4 == 0,
            let data = normalized(joined),
            let start = block.first?.range.lowerBound, let end = block.last?.range.upperBound
        else { return nil }
        return Candidate(data: data, declaredMime: nil, range: start..<end)
    }

    static func normalized(_ raw: String) -> String? {
        let stripped = raw.filter { !$0.isWhitespace }
        guard !stripped.isEmpty else { return nil }
        let body = stripped.prefix { $0 != "=" }
        guard body.allSatisfy({ $0.isASCII && ($0.isLetter || $0.isNumber || $0 == "+" || $0 == "/") })
        else { return nil }
        let pad = stripped.dropFirst(body.count)
        guard pad.allSatisfy({ $0 == "=" }), pad.count <= 2 else { return nil }
        let remainder = body.count % 4
        if remainder == 1 { return nil }
        return remainder == 0 ? String(body) : String(body) + String(repeating: "=", count: 4 - remainder)
    }
}
